src/status/condition/mod.rs

```diff
@@ -0,0 +1,472 @@
+pub mod daemonset;
+pub mod operations;
+pub mod statefulset;
+
+use chrono::Utc;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::Time;
+use schemars::{self, JsonSchema};
+use serde::{Deserialize, Serialize};
+use strum::EnumCount;
+
+/// A **data structure** that contains a vector of `ClusterCondition`s.
+/// Should usually be implemented on the status of a `CustomResource` or the `CustomResource` itself.
+pub trait HasStatusCondition {
+    fn conditions(&self) -> Vec<ClusterCondition>;
+}
+
+/// A **data structure** that produces a `ClusterConditionSet` containing all required
+/// `ClusterCondition`s.
+pub trait ConditionBuilder {
+    fn build_conditions(&self) -> ClusterConditionSet;
+}
+
+/// Computes the final conditions to be set in the operator status condition field.
+///
+/// # Arguments
+///
+/// * `resource` - A cluster resource or status implementing [`HasStatusCondition`] in order to
+///    retrieve the "current" conditions set in the cluster. This is required to  compute
+///    condition change and set proper update / transition times.
+/// * `condition_builders` - A slice of structs implementing [`ConditionBuilder`]. This can be a
+///    one of the predefined ConditionBuilders like `DaemonSetConditionBuilder` or a custom
+///    implementation for special resources or different behavior.
+///
+/// # Examples
+/// ```
+/// use stackable_operator::status::condition::daemonset::DaemonSetConditionBuilder;
+/// use stackable_operator::status::condition::statefulset::StatefulSetConditionBuilder;
+/// use k8s_openapi::api::apps::v1::{DaemonSet, StatefulSet};
+/// use stackable_operator::status::condition::{ClusterCondition, ConditionBuilder, HasStatusCondition, compute_conditions};
+///
+/// struct ClusterStatus {
+///     conditions: Vec<ClusterCondition>
+/// }
+///
+/// impl HasStatusCondition for ClusterStatus {
+///     fn conditions(&self) -> Vec<ClusterCondition> {
+///         self.conditions.clone()
+///     }
+/// }
+///
+/// let mut daemonset_condition_builder = DaemonSetConditionBuilder::default();
+/// daemonset_condition_builder.add(DaemonSet::default());
+///
+/// let mut statefulset_condition_builder = StatefulSetConditionBuilder::default();
```

Comment:
Im fine with adapting it. The thoughts behind it were `HasStatusCondition` basically indicates im a Cluster that has conditions in its status. `ConditionBuilder` should be the shared interface to compute and merge conditions of anything that implements it.
Lets see if someone screems :D

Reply:
Naming is hard and I don't have a strong opinion but implementations of `CanComputeConditions` only make sense if they indeed compute conditions. This trait was not meant to be implemented on "foreign" structs that don't have conditions but somehow have to participate in building a resource status.

Also to me, it would feel awkward if all  traits would be called `Can...` , `Has...` and so on just to follow a naming pattern.

But again, naming is hard.